lib/apiservers/service/restapi/handlers/common.go

```diff
@@ -40,71 +41,82 @@ type buildDataParams struct {
 	thumbprint      *string
 	datacenter      *string
 	computeResource *string
+	vchID           *string
 }
 
-func buildData(op trace.Operation, params buildDataParams, principal interface{}) (*data.Data, error) {
-	d := data.Data{
+func buildDataAndValidateTarget(op trace.Operation, params buildDataParams, principal interface{}) (*data.Data, *validate.Validator, error) {
+	data := &data.Data{
 		Target: &common.Target{
 			URL: &url.URL{Host: params.target},
 		},
 	}
 
 	if c, ok := principal.(Credentials); ok {
-		d.Target.User = c.user
-		d.Target.Password = &c.pass
+		data.Target.User = c.user
+		data.Target.Password = &c.pass
 	} else {
-		d.Target.CloneTicket = principal.(Session).ticket
+		data.Target.CloneTicket = principal.(Session).ticket
+	}
+
+	if err := data.HasCredentials(op); err != nil {
+		return data, nil, util.NewError(http.StatusUnauthorized, "Invalid Credentials: %s", err)
 	}
 
 	if params.thumbprint != nil {
-		d.Thumbprint = *params.thumbprint
+		data.Thumbprint = *params.thumbprint
+	}
+
+	if params.computeResource != nil {
+		data.ComputeResourcePath = *params.computeResource
 	}
 
+	if params.vchID != nil {
+		data.ID = *params.vchID
+	}
+
+	// TODO (#6032): clean this up
+	var validator *validate.Validator
 	if params.datacenter != nil {
-		validator, err := validateTarget(op, &d)
+		v, err := validate.NewValidator(op, data)
 		if err != nil {
-			return nil, util.WrapError(http.StatusInternalServerError, err)
+			return data, nil, util.NewError(http.StatusBadRequest, "Validation Error: %s", err)
 		}
 
 		datacenterManagedObjectReference := types.ManagedObjectReference{Type: "Datacenter", Value: *params.datacenter}
 
-		datacenterObject, err := validator.Session.Finder.ObjectReference(op, datacenterManagedObjectReference)
+		datacenterObject, err := v.Session.Finder.ObjectReference(op, datacenterManagedObjectReference)
 		if err != nil {
-			return nil, util.WrapError(http.StatusNotFound, err)
+			return nil, nil, util.WrapError(http.StatusNotFound, err)
 		}
 
-		d.Target.URL.Path = datacenterObject.(*object.Datacenter).InventoryPath
-	}
+		data.Target.URL.Path = datacenterObject.(*object.Datacenter).InventoryPath
 
-	if params.computeResource != nil {
-		d.ComputeResourcePath = *params.computeResource
-	}
+		// Copied from NewValidator
+		v.DatacenterPath = strings.TrimPrefix(data.Target.URL.Path, "/")
+		v.Session.DatastorePath = v.DatacenterPath
 
-	return &d, nil
-}
+		validator = v
+	} else {
+		v, err := validate.NewValidator(op, data)
+		if err != nil {
+			return data, nil, util.NewError(http.StatusBadRequest, "Validation Error: %s", err)
+		}
 
-func validateTarget(op trace.Operation, d *data.Data) (*validate.Validator, error) {
-	if err := d.HasCredentials(op); err != nil {
-		return nil, fmt.Errorf("Invalid Credentials: %s", err)
-	}
+		// If dc is not set, and multiple datacenters are available, operate on all datacenters.
+		v.AllowEmptyDC()
```

Comment:
IIRC this was added solely for the operations that don't care about a specific VCH, so `ls`, and `update`, of which only `ls` is relevant for us.

I think it may be better to constrain this path based on the rest operation - I don't have confidence that we've tested what happens if you try other operations without a datacenter defined.

Reply:
The call to `.AllowEmptyDC()` is present in the `validateTarget` method today, so it's been present for all of the testing of the API/UI.

If there are not known issues with having it here, I'd be tempted to leave it for now to avoid unknown issues with removing it (since I think we've done more testing _with_ it than we would do _without_ it between now and when we ship).